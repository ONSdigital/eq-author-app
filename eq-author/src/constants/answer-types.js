export const CHECKBOX = "Checkbox";
export const RADIO = "Radio";
export const TEXTFIELD = "TextField";
export const TEXTAREA = "TextArea";
export const CURRENCY = "Currency";
export const NUMBER = "Number";
export const PERCENTAGE = "Percentage";
export const DATE = "Date";
export const DATE_RANGE = "DateRange";
export const UNIT = "Unit";
export const DURATION = "Duration";

export const ROUTING_ANSWER_TYPES = [
  RADIO,
  CURRENCY,
  NUMBER,
  PERCENTAGE,
  CHECKBOX,
  UNIT,
];

export const RADIO_OPTION = "RadioOption";
export const CHECKBOX_OPTION = "CheckboxOption";
export const MUTUALLY_EXCLUSIVE = "MutuallyExclusive";
export const MUTUALLY_EXCLUSIVE_OPTION = "MutuallyExclusiveOption";

export default [
  CHECKBOX,
  RADIO,
  TEXTFIELD,
  TEXTAREA,
  CURRENCY,
  NUMBER,
  PERCENTAGE,
  DATE,
  DATE_RANGE,
  UNIT,
  DURATION,
];
