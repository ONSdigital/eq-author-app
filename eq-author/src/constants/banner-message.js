const bannerMessageMigrated =
  "Author has been migrated and this page will turn off shortly.";
const bannerMessageVersion = "Please use latest version of Author.";
const bannerMessageQuestions =
  "If you have any questions, please contact the Author team.";

const bannerMessages = [
  bannerMessageMigrated,
  bannerMessageVersion,
  bannerMessageQuestions,
];

export default bannerMessages;
