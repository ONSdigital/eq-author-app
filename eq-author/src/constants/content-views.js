export const ANSWER = "answer";
export const METADATA = "metadata";

export default [ANSWER, METADATA];
