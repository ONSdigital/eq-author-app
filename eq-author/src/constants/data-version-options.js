const DATA_VERSION_OPTIONS = [
  {
    value: "1",
    title: "Data version 1",
    description:
      "Enables downstream processing of older questionnaires without complex features",
  },
  {
    value: "3",
    title: "Data version 3",
    description:
      "Enables downstream processing of newer questionnaires with complex features",
  },
];

export default DATA_VERSION_OPTIONS;
