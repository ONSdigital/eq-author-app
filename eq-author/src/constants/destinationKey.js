export const EndOfQuestionnaire = "EndOfQuestionnaire";
export const NextPage = "NextPage";
export const Default = "Default";
export const EndOfCurrentSection = "EndOfCurrentSection";

export const destinationKey = {
  [EndOfQuestionnaire]: "End of questionnaire",
  [EndOfCurrentSection]: "End of current section",
  [NextPage]: "Next page",
  [Default]: "Select a destination",
};
