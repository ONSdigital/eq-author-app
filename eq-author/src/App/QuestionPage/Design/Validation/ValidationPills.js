import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { compact, isFunction } from "lodash";

import PillTabs from "components/PillTabs";

import * as entityTypes from "constants/validation-entity-types";

export const Pills = styled(PillTabs)`
  width: 5em;
`;

const PillTabContent = styled.div`
  margin-top: 3em;
  margin-bottom: 0;
  width: 100%;
`;

export const ValidationPills = ({
  entityType,
  onEntityTypeChange,
  PreviousAnswer,
  Metadata,
  Now,
  Custom
}) => (
  <Pills
    value={entityType}
    onChange={onEntityTypeChange}
    options={compact([
      isFunction(Now)
        ? {
            id: "Now",
            title: "Start date",
            render: () => (
              <PillTabContent>
                <Now />
              </PillTabContent>
            )
          }
        : null,
      isFunction(PreviousAnswer)
        ? {
            id: "PreviousAnswer",
            title: "Previous answer",
            render: () => (
              <PillTabContent>
                <PreviousAnswer />
              </PillTabContent>
            )
          }
        : null,
      isFunction(Metadata)
        ? {
            id: "Metadata",
            title: "Metadata",
            render: () => (
              <PillTabContent>
                <Metadata />
              </PillTabContent>
            )
          }
        : null,
      isFunction(Custom)
        ? {
            id: "Custom",
            title: "Custom",
            render: () => (
              <PillTabContent>
                <Custom />
              </PillTabContent>
            )
          }
        : null
    ])}
  />
);

ValidationPills.propTypes = {
  entityType: PropTypes.oneOf(Object.values(entityTypes)).isRequired,
  onEntityTypeChange: PropTypes.func.isRequired,
  PreviousAnswer: PropTypes.func,
  Metadata: PropTypes.func,
  Custom: PropTypes.func,
  Now: PropTypes.func
};
