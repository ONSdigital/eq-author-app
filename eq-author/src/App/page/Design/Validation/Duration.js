import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { colors } from "constants/theme";

import { Number, Select } from "components/Forms";
import { Grid, Column } from "components/Grid";

import { DAYS, MONTHS, YEARS } from "constants/durations";

export const DurationNumber = styled(Number)`
  width: 100%;

  ${({ hasError }) =>
    hasError &&
    `
    border-color: ${colors.red};
    outline-color: ${colors.red};
    box-shadow: 0 0 0 2px ${colors.red};
    border-radius: 4px;
    margin-bottom: 0;
  `}
`;

const UNITS = [DAYS, MONTHS, YEARS];

const Duration = ({
  name,
  units,
  duration: { value, unit },
  onChange,
  onUpdate,
  hasError,
}) => (
    <Grid>
      <Column cols={2}>
        <DurationNumber
          id={`${name}-value`}
          name={`${name}.value`}
          value={value}
          onChange={onChange}
          onBlur={onUpdate}
          max={99999}
          min={0}
          hasError={hasError}
        />
      </Column>
      <Column cols={4}>
        <Select
          id={`${name}-unit`}
          name={`${name}.unit`}
          value={unit}
          onChange={onChange}
          onBlur={onUpdate}
        >
          {units.map(unit => (
            <option key={unit} value={unit}>
              {unit.toLowerCase()}
            </option>
          ))}
        </Select>
      </Column>
    </Grid>
  );

Duration.propTypes = {
  name: PropTypes.string.isRequired,
  duration: PropTypes.shape({
    value: PropTypes.number,
    unit: PropTypes.oneOf(UNITS).isRequired,
  }).isRequired,
  units: PropTypes.arrayOf(PropTypes.string).isRequired,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  hasError: PropTypes.object, // eslint-disable-line
};

export default Duration;
