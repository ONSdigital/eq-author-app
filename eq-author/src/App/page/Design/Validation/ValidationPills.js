import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { compact, isFunction } from "lodash";

import PillTabs from "components/PillTabs";

import * as entityTypes from "constants/validation-entity-types";

export const Pills = styled(PillTabs)`
  width: 5em;
`;

const PillTabContent = styled.div`
  margin-top: 3em;
  margin-bottom: 0;
  width: 100%;
`;

export const ValidationPills = ({
  entityType,
  onEntityTypeChange,
  PreviousAnswer,
  Metadata,
  Now,
  Custom,
  ...rest
}) => {
  const handleOnChange = (value) =>
    onEntityTypeChange({ name: "entityType", value });

  return (
    <Pills
      value={entityType}
      onChange={handleOnChange}
      options={compact([
        isFunction(Now)
          ? {
              id: "Now",
              title: "Start date",
              render: () => (
                <PillTabContent>
                  <Now {...rest} />
                </PillTabContent>
              ),
            }
          : null,
        isFunction(PreviousAnswer)
          ? {
              id: "PreviousAnswer",
              title: "Previous answer",
              render: () => (
                <PillTabContent>
                  <PreviousAnswer {...rest} />
                </PillTabContent>
              ),
            }
          : null,
        isFunction(Metadata)
          ? {
              id: "Metadata",
              title: "Metadata",
              render: () => (
                <PillTabContent>
                  <Metadata {...rest} />
                </PillTabContent>
              ),
            }
          : null,
        isFunction(Custom)
          ? {
              id: "Custom",
              title: "Custom",
              render: () => (
                <PillTabContent>
                  <Custom {...rest} />
                </PillTabContent>
              ),
            }
          : null,
      ])}
    />
  );
};

ValidationPills.propTypes = {
  entityType: PropTypes.oneOf(Object.values(entityTypes)).isRequired,
  onEntityTypeChange: PropTypes.func.isRequired,
  PreviousAnswer: PropTypes.func,
  Metadata: PropTypes.func,
  Custom: PropTypes.func,
  Now: PropTypes.func,
};
