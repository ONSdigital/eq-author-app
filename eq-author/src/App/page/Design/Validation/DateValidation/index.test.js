import React from "react";
import { shallow } from "enzyme";
import { ValidationPills } from "../ValidationPills";
import { UnwrappedDateValidation } from ".";
import Duration from "../Duration";
import { DATE, DATE_RANGE } from "constants/answer-types";
import { CUSTOM } from "constants/validation-entity-types";

const createWrapper = (props, render = shallow) =>
  render(<UnwrappedDateValidation {...props} />);

describe("Date Validation", () => {
  let props, wrapper;
  let onCustomNumberValueChange = jest.fn();
  let onChangeUpdate = jest.fn();
  let onChange = jest.fn();
  let onUpdate = jest.fn();
  let onToggleValidationRule = jest.fn();

  beforeEach(() => {
    props = {
      validation: {
        id: "123",
        enabled: true,
        customDate: "2018-01-01",
        offset: {
          value: 5,
          unit: "Months",
        },
        relativePosition: "Before",
        entityType: CUSTOM,
        validationErrorInfo: {
          errors: [
            {
              errorCode: "ERR_NO_VALUE",
              field: "custom",
              id: "0efd3ed1-8e0d-4b0c-9e39-59010751dbdf",
              type: "validation",
            },
          ],
        },
      },
      answer: {
        id: "1",
        type: DATE,
        properties: {
          format: "YYYY",
        },
      },
      onCustomNumberValueChange: onCustomNumberValueChange,
      onChangeUpdate: onChangeUpdate,
      onChange: onChange,
      onUpdate: onUpdate,
      onToggleValidationRule: onToggleValidationRule,
      displayName: "foobar",
      readKey: "read",
      testId: "test-id",
    };

    wrapper = createWrapper(props);
  });

  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render previous answer, metadata, custom and now for date type answer", () => {
    const answer = {
      ...props.answer,
      type: DATE,
    };
    const wrapper = createWrapper({ ...props, answer });
    expect(wrapper.find(ValidationPills)).toMatchSnapshot();
  });

  it("should render metadata and custom for date range type answer", () => {
    const answer = {
      ...props.answer,
      type: DATE_RANGE,
    };
    const wrapper = createWrapper({ ...props, answer });
    expect(wrapper.find(ValidationPills)).toMatchSnapshot();
  });

  it("should trigger update answer validation when the duration value changes", () => {
    const duration = wrapper.find(Duration);
    duration.simulate("change", "event");
    expect(props.onChange).toHaveBeenCalledWith("event");
    duration.simulate("update", "event");
    expect(props.onUpdate).toHaveBeenCalledWith("event");
  });

  it("should correctly render 'now' entity type", () => {
    const Now = wrapper.find(ValidationPills).prop("Now");
    expect(shallow(<Now />)).toMatchSnapshot();
  });

  it("should filter the available units for the mm/yyyy", () => {
    const answer = {
      ...props.answer,
      properties: {
        format: "mm/yyyy",
      },
    };
    const wrapper = createWrapper({ ...props, answer });
    const duration = wrapper.find(Duration);
    expect(duration.prop("units")).toMatchSnapshot();
  });

  it("should filter the available units for the dd/mm/yyyy", () => {
    const answer = {
      ...props.answer,
      properties: {
        format: "dd/mm/yyyy",
      },
    };
    const wrapper = createWrapper({ ...props, answer });
    const duration = wrapper.find(Duration);
    expect(duration.prop("units")).toMatchSnapshot();
  });

  it("should filter the available units for the yyyy", () => {
    const answer = {
      ...props.answer,
      properties: {
        format: "yyyy",
      },
    };
    const wrapper = createWrapper({ ...props, answer });
    const duration = wrapper.find(Duration);
    expect(duration.prop("units")).toMatchSnapshot();
  });

  it("should render all units for date range answers", () => {
    const answer = {
      ...props.answer,
      type: DATE_RANGE,
    };
    const wrapper = createWrapper({ ...props, answer });
    const duration = wrapper.find(Duration);
    expect(duration.prop("units")).toMatchSnapshot();
  });
});
