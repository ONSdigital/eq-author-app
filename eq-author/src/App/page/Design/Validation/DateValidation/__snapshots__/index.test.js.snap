// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Validation should correctly render 'now' entity type 1`] = `
<components__StartDateText>
  The date the respondent begins the survey
</components__StartDateText>
`;

exports[`Date Validation should filter the available units for the dd/mm/yyyy 1`] = `
Array [
  "Days",
  "Months",
  "Years",
]
`;

exports[`Date Validation should filter the available units for the mm/yyyy 1`] = `
Array [
  "Months",
  "Years",
]
`;

exports[`Date Validation should filter the available units for the yyyy 1`] = `
Array [
  "Years",
]
`;

exports[`Date Validation should render 1`] = `
<div>
  <Grid
    align="top"
    fillHeight={true}
  >
    <AlignedColumn
      cols={3}
    >
      <ValidationTitle>
        foobar
         is
      </ValidationTitle>
    </AlignedColumn>
    <Column
      cols={9}
    >
      <Duration
        duration={
          Object {
            "unit": "Months",
            "value": 5,
          }
        }
        hasError={
          Object {
            "errorCode": "ERR_NO_VALUE",
            "field": "custom",
            "id": "0efd3ed1-8e0d-4b0c-9e39-59010751dbdf",
            "type": "validation",
          }
        }
        name="offset"
        onChange={[MockFunction]}
        onUpdate={[MockFunction]}
        units={
          Array [
            "Years",
          ]
        }
        value="Months"
      />
    </Column>
  </Grid>
  <Grid
    align="top"
    fillHeight={true}
  >
    <Column
      cols={3}
    >
      <components__ConnectedPath />
    </Column>
    <Column
      cols={9}
    >
      <components__StyledError>
        Value is required
      </components__StyledError>
    </Column>
  </Grid>
  <Grid
    align="top"
    fillHeight={true}
  >
    <AlignedColumn
      cols={3}
    >
      <PositionPicker
        onChange={[MockFunction]}
        onUpdate={[MockFunction]}
        value="Before"
      />
      <Component />
    </AlignedColumn>
    <Column
      cols={9}
    >
      <ValidationPills
        Custom={[Function]}
        Metadata={[Function]}
        Now={[Function]}
        PreviousAnswer={[Function]}
        answer={
          Object {
            "id": "1",
            "properties": Object {
              "format": "YYYY",
            },
            "type": "Date",
          }
        }
        entityType="Custom"
        onChange={[MockFunction]}
        onChangeUpdate={[MockFunction]}
        onEntityTypeChange={[MockFunction]}
        onUpdate={[MockFunction]}
        readKey="read"
        validation={
          Object {
            "customDate": "2018-01-01",
            "enabled": true,
            "entityType": "Custom",
            "id": "123",
            "offset": Object {
              "unit": "Months",
              "value": 5,
            },
            "relativePosition": "Before",
            "validationErrorInfo": Object {
              "errors": Array [
                Object {
                  "errorCode": "ERR_NO_VALUE",
                  "field": "custom",
                  "id": "0efd3ed1-8e0d-4b0c-9e39-59010751dbdf",
                  "type": "validation",
                },
              ],
            },
          }
        }
      />
    </Column>
  </Grid>
</div>
`;

exports[`Date Validation should render all units for date range answers 1`] = `
Array [
  "Years",
]
`;

exports[`Date Validation should render metadata and custom for date range type answer 1`] = `
<ValidationPills
  Custom={[Function]}
  Metadata={[Function]}
  answer={
    Object {
      "id": "1",
      "properties": Object {
        "format": "YYYY",
      },
      "type": "DateRange",
    }
  }
  entityType="Custom"
  onChange={[MockFunction]}
  onChangeUpdate={[MockFunction]}
  onEntityTypeChange={[MockFunction]}
  onUpdate={[MockFunction]}
  readKey="read"
  validation={
    Object {
      "customDate": "2018-01-01",
      "enabled": true,
      "entityType": "Custom",
      "id": "123",
      "offset": Object {
        "unit": "Months",
        "value": 5,
      },
      "relativePosition": "Before",
      "validationErrorInfo": Object {
        "errors": Array [
          Object {
            "errorCode": "ERR_NO_VALUE",
            "field": "custom",
            "id": "0efd3ed1-8e0d-4b0c-9e39-59010751dbdf",
            "type": "validation",
          },
        ],
      },
    }
  }
/>
`;

exports[`Date Validation should render previous answer, metadata, custom and now for date type answer 1`] = `
<ValidationPills
  Custom={[Function]}
  Metadata={[Function]}
  Now={[Function]}
  PreviousAnswer={[Function]}
  answer={
    Object {
      "id": "1",
      "properties": Object {
        "format": "YYYY",
      },
      "type": "Date",
    }
  }
  entityType="Custom"
  onChange={[MockFunction]}
  onChangeUpdate={[MockFunction]}
  onEntityTypeChange={[MockFunction]}
  onUpdate={[MockFunction]}
  readKey="read"
  validation={
    Object {
      "customDate": "2018-01-01",
      "enabled": true,
      "entityType": "Custom",
      "id": "123",
      "offset": Object {
        "unit": "Months",
        "value": 5,
      },
      "relativePosition": "Before",
      "validationErrorInfo": Object {
        "errors": Array [
          Object {
            "errorCode": "ERR_NO_VALUE",
            "field": "custom",
            "id": "0efd3ed1-8e0d-4b0c-9e39-59010751dbdf",
            "type": "validation",
          },
        ],
      },
    }
  }
/>
`;
